### ISSUE TITLE
`node_access_acquire_grants()` is invoked too late in BAckdrop

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#5591]
## Description of the bug

In D7, `node_access_acquire_grants()` was invoked by `node_save()` after invoking all the "normal" node hooks (presave, insert, update or delete).

In Backdrop, this order is inverted. In `NodeStorageController::save()` the method `postSave()` (which invokes `node_access_acquire_grants()`) is invoked BEFORE the method `invokeHooks()`. 

This causes severe issues to modules that use `node_access`, for example Organic Groups, since operations that normally should take place before acquiring node access grants for new nodes end up happening too late. For example, Organic Groups needs to write a record in the `og_membership` table to establish a relationship between a group node and a group content node, BEFORE  running `node_access_acquire_grants()`.

Since this is not happening in the right order, new group content nodes that belong to private groups end up available to everyone to see.

## Steps To Reproduce

To reproduce you need to install Organic Groups. 

But better yet, you can see this order reversal clearly by looking at D7's `node_save()` and comparing it to `NodeStorageController::save()`

## Additional information

A simple solution would be to invert the order of these two method invocation. `postSave()` currently only exists to invoke `node_access_acquire_grants()` - it doesn't do anything else.  It did not exist in D7.  A PR would be really simple but I'd like some feedback on this, please.

**Comment by indigoxela** [comment]
> In NodeStorageController::save() the method postSave() (which invokes node_access_acquire_grants()) is invoked BEFORE the method invokeHooks().

I suspect this affects all modules that work with node grants and entity insert hooks. I wonder how they solved that - if they ran into the same problem. It might be that they didn't, though.

**Comment by argiepiano** [comment]
PR https://github.com/backdrop/backdrop/pull/4049 provided.

**Comment by argiepiano** [comment]
OK, we have some failing tests. I'll check what's going on

**Comment by argiepiano** [comment]
I've discovered the issue with the failing test. 
- The test provide an implementation of hook_node_insert that modifies and resaves the node
- Since the PR swapped the order hooks, and now the insert/update hook invocations happen before the call to `postSave()` and the subsequent `node_access_acquire_grants()`, the node with the `update` operation gets to `node_access_acquire_grants()` before the node with the `insert` operation

The Backdrop code for the invocation of `node_access_acquire_grants()` differs in one significant way from the one in D7: when the operation is something other than `update`, `node_access_acquire_grants()` does not delete old grants from the `node_access` table. This causes an exception for a duplicate entry in the `node_access` table.

This isn't an issue when `node_access_acquire_grants()` is performed before the insert/update hooks in the current unpatched version Backdrop.

So, a proposed solution (if you all think the swapping is desirable - which I do in order to allow modules to act before writing grants to the table), is to always call `node_access_acquire_grants()` with a TRUE value as its second argument, so that existing grants for the node are always deleted from the DB before reinserting them. This is a bit less efficient, but it's the only way to keep the current order. This BTW is how D7 did it.

I'll provide a PR and y'all can decide if this is worth it.

**Comment by argiepiano** [comment]
All tests passing with the modified PR

**Comment by indigoxela** [comment]
This looks correct in a way, but... I feel a bit uncomfortable with changing the default value for that function. But I admit, that I don't fully understand the current logic yet.

```
      $this->postSave($entity, $op == 'update');
      $this->invokeHook($op, $entity);
```
I agree, that this order seems odd.

One comment in function postSave causes me head-scratching:

> There's no need to delete existing records if the node is new.

Nothing in function postSave checks if the node is new, though.

And also: the parameters are pretty confusing.

```
$this->postSave($entity, $op == 'update');

function postSave(EntityInterface $node, $update) {

node_access_acquire_grants($node, $update);
```
What's confusing? postSave gets called with a string parameter (the action), but postSave actually hands that over as-is to node_access_acquire_grants, which actually expects a boolean (whether to delete existing node access records before inserting or not). :confused: 

The docblock in function postSave in entity.controller.inc clearly states that it's supposed to be a boolean:

```
   * @param $update
   *   (bool) TRUE if the entity has been updated, or FALSE if it has been
   *   inserted.
```
So, if in node.entity.inc it's used like that in line 498:
```
$this->postSave($entity, $op == 'update');
```
That's clearly just wrong.

**Comment by indigoxela** [comment]
@argiepiano can we try the following:

- Still switch the order in function save
- Back to the initial param in function postSave (it should be the same as in the parent class)
- Call postSave in node.entity.inc correctly (with the right boolean)
- Update the docblock for param $update in Node::postSave

Then, if automated tests fail, let's have a closer look at the test.

**Comment by indigoxela** [comment]
An hour later... switching the order of postSave and invokeHook seems to break something with hook_node_insert.

The failing test (when switching the order) has been created for this [Drupal issue](https://www.drupal.org/project/drupal/issues/1146244) - and this problem seems to be back when the order is switched.

**Comment by indigoxela** [comment]
If the order is switched, the $op variable can be outdated, because hook_node_insert may already have saved the node.

If I switch, but also change the param, the test passes:
```
      $this->invokeHook($op, $entity);
      $this->postSave($entity, TRUE);
```

The node might not be new anymore, because some module could have saved the node in hook_node_insert.

The big question now: is it OK to switch the order? Or is there a reliable way to check if the node is still new at that point? Note that the node property is_new has been unset above, so isNew() wouldn't work anymore. The only option I see is to always set the param for postSave (and node_access_acquire_grants) to TRUE - which seems odd.

Also: shouldn't the first node save actually be done by the node module, not any module that implements hook_node_insert?

**Comment by argiepiano** [comment]
Hi @indigoxela  - thanks for all the messages. I think you found out that the call to postSave actually did use a boolean by using `==`. That threw me for a loop for a little while.

I think you also found why the `update` op reaches postSave before the `insert` op, given the node_save() operation being performed within that hook_node_insert().

I have to say that, like you,  I find this odd (saving the node again within a hook_node_insert), but there are cases I've seen this done.  

>Also: shouldn't the first node save actually be done by the node module, not any module that implements hook_node_insert?

The call to `node_save()` within `node_test_node_insert()` actually ends up calling the node module to do the saving. But I may be misunderstanding what you meant above.

Your code: `$this->postSave($entity, TRUE);`  - that's more or less what I did in my pr. I left the second argument blank, and created a TRUE default for it in `postSave`. The name of the parameter in `postSave` is confusing (`$update`). That parameter, as you noticed, is a boolean, and basically tells `node_access_acquire_grants` to delete any entry for that node in the `node_access` table before rewriting.

> One comment in function postSave causes me head-scratching:
>> There's no need to delete existing records if the node is new.

> Nothing in function postSave checks if the node is new, though.
  
Yup, that caused me to pause too. What that comment means is that (with the existing Backdrop code)  the second parameter  is false if the node op was `insert`. The idea behind this was that you did not need to bother removing ("deleting") the entry in the `node_access` table if you are saving the node for the very first time. But as we have seen, there are cases in which the `update` op gets to `postSave` BEFORE the `insert` op for the same module (which is obviously poor design). So my PR and the one you posted from the Drupal queue mitigated that issue.

**Comment by argiepiano** [comment]
>Or is there a reliable way to check if the node is still new at that point?

One more comment. 
I did consider checking the node's `is_new` to decide whether to call `node_access_acquire_grants` or not. I believe that we still need to call `node_access_acquire_grants` when we update an existing node - there may be cases where the grants may change based on the new values of the updated node.

So, the issue here is not whether we should call `node_access_acquire_grants` or not. Rather, it's whether we can ever call `node_access_acquire_grants` with FALSE as it's second argument at this point. I argue we should always call it with TRUE, as we always need to delete any existing grants before creating new ones (and avoid the exception of trying to write `node_access` with an already existing record)

**Comment by docwilmot** [comment]
Dont known much about node grants but the docblock for `EntityDatabaseStorageController::postSave()`, which `NodeStorageController()` extends, says:
> Used after the entity is saved, but before invoking the insert or update hook.

I dont think we should reverse the execution order just for Node then; that would lead to issues in the future if Node behaves differently to other Entities. Could we either add the call to `node_access_acquire_grants()` in `invokeHook()` itself after the update/insert calls, OR directly in `NodeStorageController::save()` (after the invokes) as D7 does?

**Comment by indigoxela** [comment]
> I dont think we should reverse the execution order just for Node then; that would lead to issues in the future if Node behaves differently to other Entities.

That's a valid point.

> Could we either add the call to node_access_acquire_grants() in invokeHook() itself after the update/insert calls, OR directly in NodeStorageController::save() (after the invokes) as D7 does?

Moving node_access_acquire_grants() to save() means, that postSave() wouldn't do anything.
I wish, I'd understand how this all is supposed to work. (Still failing to do so :wink:)

Currently it's node_access_acquire_grants() is invoked too early - do we agree on that? Are there possible side effects when we move it to save? Any concerns re that?

**Comment by docwilmot** [comment]
> Moving node_access_acquire_grants() to save() means, that postSave() wouldn't do anything.

We could just delete `PostSave()` since its still in the parent class and it does nothing else in Node. I suspect `node_access_acquire_grants()` would be fine in `save()`.

Brief words, late for work so may not have thought this through fully.

**Comment by argiepiano** [comment]
@docwilmot, @indigoxela, I've modified the PR following @docwilmot suggestions:

- I moved the call to `node_access_acquire_grants()` to the `save()` method, and placed it after the hook invocations. This follows the same pattern as D7
- I removed the method `postSave()`, as it doesn't do anything now that  `node_access_acquire_grants` is in `save

This last one shouldn't cause any issues, since `postSave()` exists in the parent class `EntityDatabaseStorageController`. 

In the call to `node_access_acquire_grants()` I have removed the second argument. This results in the `node_access` table being rebuilt on every save, regardless of whether it's an update or an insert op. The reason why this is necessary is that, when `node_access_acquire_grants` is moved after the hook invocation, at times the call to `node_access_acquire_grants` will happen with op `update` BEFORE  `insert` on the same page request and with the same node. The old code skipped rebuilding `node_access` on insert - which can result in a fatal error when trying to write a record into `node_access` that already exists. 

@indigoxela , the reversal of the node op happens if an implementation of hook_node_insert() saves the node. I've created a very crude graph that may clarify how this happens.

![node1](https://user-images.githubusercontent.com/9938978/166089079-ff5c9da4-b9ee-4d0b-b208-01837b9cc7b8.jpg)

**Comment by indigoxela** [comment]
Odd, I just took a look at the order in EntityDatabaseStorageController:save():

```
      $this->preSave($entity);
      $this->invokeHook('presave', $entity);

      if (!$entity->isNew()) {
        $return = backdrop_write_record($this->entityInfo['base table'], $entity, $this->idKey);
        $this->resetCache(array($entity->{$this->idKey}));
        $this->postSave($entity, TRUE);
        $this->invokeHook('update', $entity);
      }
      else {
        $return = backdrop_write_record($this->entityInfo['base table'], $entity);
        // Reset general caches, but keep caches specific to certain entities.
        $this->resetCache(array());
        $this->postSave($entity, FALSE);
        $this->invokeHook('insert', $entity);
      }
```
So in the parent class the order is:
```
$this->preSave
$this->invokeHook
$this->postSave
```
So it was actually the node module, that switched the original order. If the node module follows the order of the parent class, we'd only have to take care of the update/insert problem.

@docwilmot any thoughts on this? @argiepiano sorry for the forward and back - should have taken a look at the parent class before.

**Comment by argiepiano** [comment]
> ```
> 
> ```

> So in the parent class the order is:
> 
> 
> ```
> $this->preSave
> $this->invokeHook
> $this->postSave
> ```

Unless I'm misunderstanding what you are saying @indigoxela, this is not the case. You pasted the code above that clearly shows pastSave() happening before invokeHook(). Mind clarifying what you meant?

**Comment by indigoxela** [comment]
> You pasted the code above that clearly shows pastSave() happening before invokeHook()

Garbage, you're right, it seems like I mixed up the invokeHooks. :worm:

**Comment by argiepiano** [comment]
@docwilmot, @indigoxela, I've made the suggested changes. Any additional thoughts? Is this fix something we eventually want in core?

**Comment by indigoxela** [comment]
> Any additional thoughts?

Nothing so far, the code looks great now. :+1: 

> Is this fix something we eventually want in core?

IMO yes. Otherwise it behaves too different compared to D7. But I'm not the one to decide.

(And I still find the order of hooks in the core Entity module odd - running postSave *before* the insert/update hook, but we probably can't change that.)

@argiepiano what's the easiest way to test this?

**Comment by argiepiano** [comment]
> argiepiano what's the easiest way to test this?

The easy, but imperfect way: download the contrib module `examples` and enable `example_node_access`.  Then add a `backdrop_set_message('Created an entry in node_access');` in line 362 of `node_access_example_node_grants()`, right above `return $grants;`. 

Then create a node and mark it as private. You will see the message indicating the entry in node_access was created first ("Created an entry..."), and then a message generated by `node_access_example_node_insert()` line 431 saying that "New node @nid was created and private=@private". After applying the patch here, the order of those two messages will be reverted, with the "New node" displayed first, and the "Created an entry" second.    

The harder way to test:

First a quick background: the reason for this fix was that OG, in a nutshell, is expected to create an entry in `og_membership` when a node is inserted (during an implementation of hook_node_insert), and then, after that entry is created, is expected to create an access record in `node_access` by implementing `hook_node_access_records`. But since Backdrop core has inverted the order of those two hooks, the node access record is wrongly created before the entry in og_membership, resulting in an access record with gid = 0 (meaning it's accessible by anyone).

Since I wasn't sure if this Backdrop issue would ever get merged or fixed, I went ahead and patched OG. You can see that [PR here](https://github.com/backdrop-contrib/og/pull/65). That patch is already merged in OG.

SO, you could "unpatch" OG and try the following:

1. Create a private group
2. Create a OG content node
3. Check the table content_access and find the node - you'll see that the gid (grant ID) column for it is 0, meaning it's public.

**Comment by indigoxela** [comment]
I tried the easy way first:

Two webs, one is 1.22.0-preview, the other one has the core from your branch (1.x-issue-5591).

In both instances I installed the example module and enabled node_access_example.
Only on 1.22.0-preview I was prompted to "Rebuild node permissions" after enabling the module.

In both webs I created a new post and marked it private, in both the order of messages was the same:

```
    New node XX was created and private=1
    Post A private node has been created.
```
In both webs the node was private after first save.

@argiepiano I'd suggest that you rebase your PR (#5589 should be in for better comparison).

So I suspect, the example module is no good way to test the PR changes. Is that correct?

**Comment by indigoxela** [comment]
Now for the OG testing:

Same setup as above.

Installed OG and OG Access on both sites, unpatched the workaround from OG, rebuilt permissions on both, added a Group and a Group content type, added Group visibility field to Group type, created a private group. Added group content to that group

On 1.22.0-preview:

- Private group was private
- Content assigned to private group was NOT private (after first save)

Saving that group content a second time (without changes) made it private.


On PR branch:

- Private group was private
- Content assigned to private group WAS private (on first save)

**Comment by argiepiano** [comment]
> So I suspect, the example module is no good way to test the PR changes. Is that correct?

The example module is only good to test the order of the hooks, not the privacy of the node. In the case of the example module, the node will be marked as private regardless of this PR, because it doesn't rely on the order of the hooks to mark privacy, like OG does. That's why I called this method "imperfect". It only shows the change of order, nothing else.

**Comment by argiepiano** [comment]
> Now for the OG testing:

Good. This is what's expected. So, without the PR, you must save the node twice for the privacy to take effect. With this PR, the first save works as expected. That's exactly the behavior this PR solves.

Is this working for you then?

**Comment by argiepiano** [comment]
May I get a milestone please?

**Comment by bugfolder** [comment]
Relying on others having done the testing on this. Reviewed the code, LGTM.

**Comment by argiepiano** [comment]
@bugfolder the original problem with OG that prompted this issue was solved in a different way within OG. I don't think there is any immediate reason to make this change - so, if it ain't broke... perhaps we shouldn't.

**Comment by bugfolder** [comment]
@argiepiano, fine by me, I'm just trying to clear some of the PRs that are in purgatory (and within my humble abilities to address). If we're not going to make this change, then should be the issue be closed? Or would the issue be likely to affect other (or future) contrib modules?

**Comment by argiepiano** [comment]
To summarize all this discussion:  unlike D7, `node_access_acquire_grants()` is called before the entity update/insert hooks are invoked. This difference in order hasn't created any issues in Backdrop core or with existing contribs AFAIK. It did create a serious, hard-to-crack issue with OG, since OG creates some access records within the insert/update hook calls, and those access records were not picked up by `node_access_acquire_grants` since it was called before those hooks. The solution was to call `node_access_acquire_grants` **again** within those hook implementations after creating those OG access records.  Redundant and a bit wasteful, but it was the only way. 

In a nutshell: there may be another access contrib module out there that may suffer from this change in order, who knows. But changing the order back creates the risk of breaking something in core or existing contrib. So, personally, I would not merge this, nor would I close this issue, just in case a developer in the future finds some answers to very hard-to-crack problems with ported contrib permission modules.

**Comment by bugfolder** [comment]
> So, personally, I would not merge this, nor would I close this issue...

OK. Presumably, though, its labels should change. No "PR - RTBC,", no milestone, and add "status — works as designed"?

**Comment by argiepiano** [comment]
Sounds good to me
